crates/bevy_ecs/src/schedule/stepping.rs

```diff
@@ -0,0 +1,1305 @@
+                );
+                self.cursor = Cursor {
+                    schedule: index,
+                    system: self.first_system_index_for_schedule(index),
+                }
+            }
+        }
+
+        Some(list)
+    }
+}
+
+#[derive(Default)]
+struct ScheduleState {
+    /// per-system [`SystemBehavior`]
+    behaviors: HashMap<NodeId, SystemBehavior>,
+
+    /// order of NodeIds in the schedule
+    node_ids: Vec<NodeId>,
+
+    /// changes to system behavior that should be applied the next time
+    /// [`ScheduleState::progress()`] is called
+    updates: HashMap<TypeId, Option<SystemBehavior>>,
+
+    /// This field contains the first steppable system in the schedule.
+    first: Option<usize>,
+}
+
+impl ScheduleState {
+    // set the stepping behavior for a system in this schedule
+    fn set_behavior(&mut self, system: SystemIdentifier, behavior: SystemBehavior) {
+        self.first = None;
+        match system {
+            SystemIdentifier::Node(node_id) => {
+                self.behaviors.insert(node_id, behavior);
+            }
+            SystemIdentifier::Type(type_id) => {
+                self.updates.insert(type_id, Some(behavior));
+            }
+        }
+    }
+
+    // clear the stepping behavior for a system in this schedule
+    fn clear_behavior(&mut self, system: SystemIdentifier) {
+        self.first = None;
+        match system {
+            SystemIdentifier::Node(node_id) => {
+                self.behaviors.remove(&node_id);
+            }
+            SystemIdentifier::Type(type_id) => {
+                self.updates.insert(type_id, None);
+            }
+        }
+    }
+
+    // clear all system behaviors
+    fn clear_behaviors(&mut self) {
+        self.behaviors.clear();
+        self.first = None;
+    }
+
+    // apply system behavior updates by looking up the node id of the system in
+    // the schedule, and updating `systems`
+    fn apply_updates(&mut self, schedule: &Schedule) {
+        for (node_id, system) in schedule.systems().unwrap() {
+            let behavior = self.updates.get(&system.type_id());
+            match behavior {
+                None => continue,
+                Some(None) => {
+                    self.behaviors.remove(&node_id);
+                }
+                Some(Some(behavior)) => {
+                    self.behaviors.insert(node_id, *behavior);
+                }
+            }
+        }
+        self.updates.clear();
+
+        println!("apply_updates(): {:?}", self.behaviors);
+    }
+
+    fn skipped_systems(
+        &mut self,
+        schedule: &Schedule,
+        start: usize,
+        mut action: Action,
+    ) -> (FixedBitSet, Option<usize>) {
+        use std::cmp::Ordering;
+
+        // if our NodeId list hasn't been populated, copy it over from the
+        // schedule
+        if self.node_ids.len() != schedule.systems_len() {
+            self.node_ids = schedule.executable().system_ids.clone();
+        }
```

Comment:
### Reaching into SystemExecutable for NodeIds
This feels ugly to me, but I don't have a better fast approach for getting this list.  I'm avoiding iterating through `Schedule::systems()` to extract NodeIds because that's much slower.

Suggestions are welcomed here.

### Improved Schedule update detection
Right now this "works", but it relies on the fact that systems can only be added to schedules.

There's an alternate approach here where we update `Schedule` to have a `generation` or `update_at` field, and we detect changes from that.  This could resolve the other comment about `systems_len()`.

This would look like:
```rust
if schedule.generation > self.last_schedule_update {
    self.node_ids = schedule.executable().system_ids.clone();
    self.last_schedule_update = schedule.generation;
}
```

I won't make this change without a reviewer requesting it.